File: python/swapSinglyLinkedList.py
# This problem was asked by Google.

# Given the head of a singly linked list, swap every two nodes and return its head.

# For example, given 1 -> 2 -> 3 -> 4, return 2 -> 1 -> 4 -> 3.


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def swap_pairs(head):
    # Create a dummy node that acts as the previous node of the head node
    dummy = ListNode(0)
    dummy.next = head
    prev = dummy

    while prev.next and prev.next.next:
        # Identify the nodes to be swapped
        first = prev.next
        second = prev.next.next

        # Swapping the nodes
        first.next = second.next
        second.next = first
        prev.next = second

        # Move the prev pointer two nodes ahead
        prev = first

    return dummy.next

# Helper function to print the linked list
def print_list(head):
    while head:
        print(head.val, end=" -> " if head.next else "\n")
        head = head.next

# Example usage
head = ListNode(1, ListNode(2, ListNode(3, ListNode(4))))
print("Original list:")
print_list(head)

swapped_head = swap_pairs(head)
print("Swapped list:")
print_list(swapped_head)

File: go/swapSinglyLinkedList.go
// This problem was asked by Google.

// Given the head of a singly linked list, swap every two nodes and return its head.

// For example, given 1 -> 2 -> 3 -> 4, return 2 -> 1 -> 4 -> 3.

package main

import "fmt"

// ListNode represents a node in the singly linked list
type ListNode struct {
	Val  int
	Next *ListNode
}

// swapPairs swaps every two adjacent nodes and returns the new head of the list
func swapPairs(head *ListNode) *ListNode {
	// Create a dummy node to simplify edge cases
	dummy := &ListNode{Next: head}
	prev := dummy

	for prev.Next != nil && prev.Next.Next != nil {
		// Identify the nodes to be swapped
		first := prev.Next
		second := prev.Next.Next

		// Perform the swap
		first.Next = second.Next
		second.Next = first
		prev.Next = second

		// Move the prev pointer two nodes ahead
		prev = first
	}

	return dummy.Next
}

// printList is a helper function to print the linked list
func printList(head *ListNode) {
	for head != nil {
		fmt.Printf("%d -> ", head.Val)
		head = head.Next
	}
	fmt.Println("nil")
}

func main() {
	// Example usage
	head := &ListNode{1, &ListNode{2, &ListNode{3, &ListNode{4, nil}}}}
	fmt.Println("Original list:")
	printList(head)

	swappedHead := swapPairs(head)
	fmt.Println("Swapped list:")
	printList(swappedHead)
}
